codex/sales/slotqueue.nim

```diff
@@ -0,0 +1,257 @@
+  self.queue.contains(sqi)
+
+proc pop*(self: SlotQueue): Future[?!SlotQueueItem] {.async.} =
+  try:
+    let sqi = await self.queue.pop()
+    return success(sqi)
+  except CatchableError as err:
+    return failure(err)
+
+proc push*(self: SlotQueue, sqi: SlotQueueItem): ?!void =
+  if self.contains(sqi):
+    let err = newException(SlotQueueItemExistsError, "item already exists")
+    return failure(err)
+
+  if err =? self.queue.pushNoWait(sqi).mapFailure.errorOption:
+    return failure(err)
+
+  if self.queue.full():
+    # delete the last item
+    self.queue.del(self.queue.size - 1)
+
+  doAssert self.queue.len <= self.queue.size
+  return success()
+
+proc push*(self: SlotQueue, sqis: seq[SlotQueueItem]): ?!void =
+  for sqi in sqis:
+    if err =? self.push(sqi).errorOption:
+      return failure(err)
+
+  return success()
+
+proc findByRequest(self: SlotQueue, requestId: RequestId): seq[SlotQueueItem] =
+  var sqis: seq[SlotQueueItem] = @[]
+  for sqi in self.queue.items:
+    if sqi.requestId == requestId:
+      sqis.add sqi
+  return sqis
+
+proc delete*(self: SlotQueue, sqi: SlotQueueItem) =
+  self.queue.delete(sqi)
+
+proc delete*(self: SlotQueue, requestId: RequestId, slotIndex: uint64) =
+  let sqi = SlotQueueItem(requestId: requestId, slotIndex: slotIndex)
+  self.delete(sqi)
+
+proc delete*(self: SlotQueue, requestId: RequestId) =
+  let sqis = self.findByRequest(requestId)
+  for sqi in sqis:
+    self.delete(sqi)
+
+proc get*(self: SlotQueue, requestId: RequestId, slotIndex: uint64): ?!SlotQueueItem =
+  let sqi = SlotQueueItem(requestId: requestId, slotIndex: slotIndex)
+  let idx = self.queue.find(sqi)
+  if idx == -1:
+    let err = newException(SlotQueueItemNotExistsError, "item does not exist")
+    return failure(err)
+
+  return success(self.queue[idx])
+
+proc `[]`*(self: SlotQueue, i: Natural): SlotQueueItem =
+  self.queue[i]
+
+proc dispatch(self: SlotQueue,
+              worker: SlotQueueWorker,
+              sqi: SlotQueueItem) {.async.} =
+
+
+
+  if onProcessSlot =? self.onProcessSlot:
+    await onProcessSlot(sqi, worker.processing)
+    await worker.processing
+
+  self.workers.keepItIf(it != worker)
+
+proc start*(self: SlotQueue) {.async.} =
+  if self.running:
+    return
+
+  self.running = true
+
+  proc handleErrors(udata: pointer) {.gcsafe.} =
+    var fut = cast[FutureBase](udata)
+    if fut.failed() and not fut.error.isNil:
+      error "slot queue error encountered during processing",
+        error = fut.error.msg
+
+  while self.running: # and self.workers.len.uint < self.maxWorkers:
+    if self.workers.len.uint >= self.maxWorkers:
+      await sleepAsync(1.millis)
+      continue
+
+    try:
+      self.next = self.queue.pop()
+      self.next.addCallback(handleErrors)
+      let sqi = await self.next # if queue empty, should wait here for new items
+      let worker = SlotQueueWorker.new()
+      self.workers.add worker
+      asyncSpawn self.dispatch(worker, sqi)
```

Comment:
Using `asyncSpawn` here means that any error that is raised in `dispatch` cause the node to crash

Reply:
`dispatch` should now be fully covered in error handling.